### PR TITLE
Implement v2 author endpoints

### PR DESCRIPTION
**Pull request by babygoat**
- [x] `/v2/author`
- [x] `/v2/author/:author_id`
- [x] `/v2/author/:author_id/post`

### CODE REVIEW

**Comment by babygoat** [comment]
Updated. 
Since 9aa2aa2 , the endpoints are ready for integration tests.

TODO items:
- [ ] Integrate with new agolia  records
- [ ] Configure with proper timeout value

These could be addressed in the following PR(different).

**Comment by taylrj** [comment]
**Issue description**

When creating `GET` request to `http://localhost:8080/v2/authors?keywords=&limit=24&offset=0`, get `thumbnail` attribute in null value as the response.

According to the [doc](https://github.com/twreporter/go-api/blob/398d5d4bba04af3e482e14cc51faaf3ef3aa90fe/doc/news/authors.apib#L108), we expect the value of `thumbnail` is in `image` type.

**Steps to Reproduce**

1. Run `twreporter-dev-env up -d` in terminal for environment setting
2. go run main.go
3. Send request to `http://localhost:8080/v2/authors?keywords=&limit=24&offset=0` and get the response

image

**Comment by taylrj** [comment]
**Issue Description**

When creating GET request to `http://localhost:8080/v2/authors/571de7bbdae62379576d7f42/posts`, receive 504 status code as the response.

By utilizing the same user id above (`571de7bbdae62379576d7f42`) with `v1` endpoint (http://localhost:8080/v1/search/posts?keywords=571de7bbdae62379576d7f42&hitsPerPage=5&page=1), which responds with 486 author posts.

**Steps to Reproduce**

1. Run `twreporter-dev-env up -d` in terminal for environment setting
2. go run main.go
3. Send `GET` request to `http://localhost:8080/v2/authors/571de7bbdae62379576d7f42/posts` and get the response

image

**Comment by babygoat** [comment]
> **Issue Description**
> 
> When creating GET request to `http://localhost:8080/v2/authors/571de7bbdae62379576d7f42/posts`, receive 504 status code as the response.
> 
> By utilizing the same user id above (`571de7bbdae62379576d7f42`) with `v1` endpoint (http://localhost:8080/v1/search/posts?keywords=571de7bbdae62379576d7f42&hitsPerPage=5&page=1), which responds with 486 author posts.
> 
> **Steps to Reproduce**
> 
>     1. Run `twreporter-dev-env up -d` in terminal for environment setting
> 
>     2. go run main.go
> 
>     3. Send `GET` request to `http://localhost:8080/v2/authors/571de7bbdae62379576d7f42/posts` and get the response
> 
> 
> image

How do you restore the data for testing? It seems to me that the query does not utilize index or something unexpectedly happened during testing. My attempts is around 1s for (non-cached) query and 1xxms for (cached) query.

**Comment by babygoat** [comment]
> **Issue Description**
> 
> When creating `GET` request to `http://localhost:8080/v2/authors?keywords=&limit=24&offset=0`, get `thumbnail` attribute in null value in the response.
> 
> According to the [doc](https://github.com/twreporter/go-api/blob/398d5d4bba04af3e482e14cc51faaf3ef3aa90fe/doc/news/authors.apib#L108), we expect the value of `thumbnail` is in `image` type.
> 
> **Steps to Reproduce**
> 
>     1. Run `twreporter-dev-env up -d` in terminal for environment setting
> 
>     2. go run main.go
> 
>     3. Send request to `http://localhost:8080/v2/authors?keywords=&limit=24&offset=0` and get the response
> 
> 
> image

Thanks for reporting. I mocked the record with incorrect field name. Address in 6f9abef

**Comment by taylrj** [comment]
> How did you restore the data for testing? It seemed to me that the query did not utilize index or something unexpectedly happened during the testing. My attempts was around 1s for (non-cached) query and 1xxms for (cached) query.

After digging out a bit, we found out that this is caused by a local file `config.yaml` which is outdated and does not contains timeout setting.

In order to avoid suffering outdated config issue, I'll stop maintaining `config.yaml` file and utilize `config.go` directly.

File: tests/pre_test_environment_setup.go [hunk 1]
```diff
@@ -181,6 +185,9 @@ func setUpDBEnvironment() (*gorm.DB, *mgo.Session, *mongodriver.Client) {
 	// set up collections in mongoDB
 	runMgoMigration(mgoDB)
 
+	// cleanup collections in mongo testdb (used by mongo-go-driver)
+	client.Database(testMongoDB).Drop(context.Background())
+
```
Comment (taylrj) [inline tests/pre_test_environment_setup.go:190]: Q: Why does it cleanup collections when the database is just being created ?

File: internal/news/mongo.go [hunk 2]
```diff
@@ -119,6 +124,21 @@ func (mf mongoFilter) BuildElements() []bson.E {
...
 			}
+		case authorFilter:
+			v := fieldV.Interface().(authorFilter)
+			if v.ID != "" {
+				var id interface{} = v.ID
+				objectID, err := primitive.ObjectIDFromHex(v.ID)
+				if err == nil {
+					id = objectID
+				}
+				elements = append(elements, bson.E{Key: mongo.OpOr, Value: bson.A{
```
[the remark this answers is not in the document]
Comment (babygoat) [reply internal/news/mongo.go]: I'm considering removing this utility function as this does not increase the readability than `bson.E`

File: tests/pre_test_environment_setup.go (hunk 1 above)
Comment (babygoat) [reply tests/pre_test_environment_setup.go:190]: What if we want to restart failure tests with records left behind in the previous round?

**Comment by taylrj** [comment]
LGTM

**Comment by babygoat** [comment]
Updated. If no further concern, I will rebase and merge it and make a new PR with agolia query.

**Comment by taylrj** [comment]
Sure, current change looks good to me.

**Pull request merged**